### ISSUE TITLE
Publish informative events about lifecycle from most important models using AMQP

### ISSUE THREAD

**Issue by yivo** [openware/peatio#757]
Deposit submitted
```javascript
{
  iss: "peatio",
  jti: SecureRandom.uuid, or cache_key
  iat: 1241235231,
  exp: 1251235543,
  data: {
   event: {
     type: "model",
     name: "deposit.update",
     record: { state: 'submitted', id: 1 },
     changes: { state: 'created' }
   }
  }
}
```

Hot wallet balance warning
```javascript
{
  iss: "peatio",
  jti: SecureRandom.uuid, or cache_key
  iat: 1241235231,
  exp: 1251235543,
  data: {
   event: {
     type: "base",
     name: "system.low_hot_wallet_balance",
     attributes: {
       balance: '100'
     }  
   }
  }
}
```

**Comment by yivo** [comment]
The small snippet demonstrating work with event API message:

```ruby
require "jwt-multisig"
require "securerandom"

jwt_payload = {
  iss: "peatio",
  jti: SecureRandom.uuid,
  data: {
    event_name: "deposit.updated",
    event_data: {
      tid:      "TID2A14A6D8CA",
      uid:      "ID2DE642BB1C",
      currency: "BTC",
      amount:   "0.38",
      state:    "accepted"
    }
  }
}

require "openssl"
private_key = OpenSSL::PKey::RSA.generate(2048)
public_key  = private_key.public_key

generated_jwt = JWT::Multisig.generate_jwt(jwt_payload, { peatio: private_key }, { peatio: 'RS256' })

Kernel.puts "GENERATED JWT", generated_jwt.to_json, "\n"

verification_result = JWT::Multisig.verify_jwt generated_jwt.deep_stringify_keys, \
  { peatio: public_key }, { verify_iss: true, iss: 'peatio', verify_jti: true }

decoded_jwt_payload = verification_result[:payload]

Kernel.puts "MATCH AFTER VERIFICATION: #{jwt_payload == decoded_jwt_payload}."
```

**Comment by mod** [comment]
Here is my suggestion:
```javascript
{
  iss: "peatio",
  jti: SecureRandom.uuid,
  iat: 1241235231,
  exp: 1251235543,
  event: "deposit.update",
  source: {
      uid:      "ID2DE642BB1C",
      tid:      "TID2A14A6D8CA",
      currency: "BTC",
      amount:   "0.38",
      state:    "prepared"
    },
  target: {
      uid:      "ID2DE642BB1C",
      tid:      "TID2A14A6D8CA",
      currency: "BTC",
      amount:   "0.38",
      state:    "accepted"
    }
  }
}
```

Events:

moduel.CRUD
- create
- update
- delete

We publish the state of the object before update and after saving was successful.

**Comment by yivo** [comment]
@mod having the data before and data after is near impossible and has the next disadvantages:
1) It is fine for events like «update», «renew». What about events which are not related to models? Logging, failures, trading events and so on. The message you suggested is not universal. 
2) It is very hard to build such payloads. In Peatio we are going to publish this type of events outside of «save-commit» hell of ActiveRecord, in some async worker which can access final data of the model. It doesn't make much sense for consumers to have this feature and I have almost never seen this similar APIs, ever in JS SPA apps where it is simple ot handle.

**Comment by yivo** [comment]
`We publish the state of the object before update and after saving was successful.`

This is not going to be so easy. If we will do it quickly we will publish lot of outdated data or ever not commited.

**Comment by yivo** [comment]
Snippet of code from chat:

```
q1 = ch.queue("", :auto_delete => true).bind(x, :routing_key => "resize")
q1.subscribe do |delivery_info, properties, payload|
  puts "[consumer] #{q1.name} received a 'resize' message"
end
```

**Comment by calj** [comment]
We would have an exchange per model, for example "events.accounts"

We need to specify the messages to ensure we can cover the following business cases:

- email verification (from barong, token, callback URL)
- login notification (date, IP, geo-location ?)
- 2FA enabled
- deposit received (amount, currency, status)
- withdrawal completed (amount, currency, status)
- trade completed
- password reset
- KyC approved
- Transaction limit reached notification
- Transaction limit changed

**Comment by yivo** [comment]
Consuming events:

```ruby
require "bunny"

Bunny.run host: "localhost", port: 5672, username: "guest", password: "guest" do |session|
  channel  = session.channel
  exchange = channel.direct("peatio.events")
  queue    = channel.queue("", auto_delete: true, durable: true, exclusive: true)
                    .bind(exchange, routing_key: "deposit.update")
  queue.subscribe manual_ack: true do |delivery_info, metadata, payload|
    Kernel.puts payload
    channel.ack(delivery_info.delivery_tag)
  rescue => e
    channel.nack(delivery_info.delivery_tag, false, true)
  end
  channel.work_pool.join
end
```

```
rabbitmqadmin publish routing_key=deposit.update payload={} exchange=peatio.events
```

**Comment by yivo** [comment]
Publishing events:
```ruby
require "bunny"

Bunny.run host: "localhost", port: 5672, username: "guest", password: "guest" do |session|
  channel  = session.channel
  exchange = channel.direct("peatio.events")
  exchange.publish("{}", routing_key: "deposit.update")
end
```

**Comment by yivo** [comment]
# `DRAFT` RabbitMQ Peatio Event API

## Overview of RabbitMQ details

Peatio submits all events into single exchange called `peatio.events`.

The routing key (same as event name) consists of two parts: 

  1) category of events or subject, like `deposit`, `withdraw`, `system`.

  2) event name, like `created`, `updated`, `destroyed`, `low_hot_wallet_balance`.

## Overview of RabbitMQ message

Each produced message in `Event API` is JWT (complete format).

This is very similar to `Management API`.

The example below demonstrates both generation and verification of JWT:

```ruby
require "jwt-multisig"
require "securerandom"

jwt_payload = {
  iss:  "peatio",
  jti:  SecureRandom.uuid,
  iat:  Time.now.to_i,
  exp:  Time.now.to_i + 60,
  data: {}
}

require "openssl"
private_key = OpenSSL::PKey::RSA.generate(2048)
public_key  = private_key.public_key

generated_jwt = JWT::Multisig.generate_jwt(jwt_payload, { peatio: private_key }, { peatio: "RS256" })

Kernel.puts "GENERATED JWT", generated_jwt.to_json, "\n"

verification_result = JWT::Multisig.verify_jwt generated_jwt.deep_stringify_keys, \
  { peatio: public_key }, { verify_iss: true, iss: "peatio", verify_jti: true }

decoded_jwt_payload = verification_result[:payload]

Kernel.puts "MATCH AFTER VERIFICATION: #{jwt_payload == decoded_jwt_payload}."
```

The RabbitMQ message is stored in JWT field called `data`.

## Overview of Event API message

The typical event looks like (JSON):

```ruby
event: {
  name: "deposit.updated",
  foo:  "...",
  bar:  "...",
  qux:  "..."
}
```

The field `event[:name]` contains event name (same as routing key).
The fields `foo`, `bar`, `qux` (just for example) are fields which carry useful information.

## Format of `deposit.created` event

```ruby
event: {
  name: "deposit.created",
  record: {
    tid:                      "TID9493F6CD41",
    uid:                      "ID092B2AF8E87",
    currency:                 "btc",
    amount:                   "0.0855",
    state:                    "submitted",
    created_at:               "2018-04-12T17:16:06+03:00",
    updated_at:               "2018-04-12T17:16:06+03:00",
    completed_at:             nil,
    blockchain_address:       "n1Ytj6Hy57YpfueA2vtmnwJQs583bpYn7W",
    blockchain_txid:          "c37ae1677c4c989dbde9ac22be1f3ff3ac67ed24732a9fa8c9258fdff0232d72",
    blockchain_confirmations: 1
  }
}
```

| Field      | Description                         |
| ---------- | ----------------------------------- |
| `record`   | The up-to-date deposit attributes.  |

## Format of `deposit.updated` event

```ruby
event: {
  name: "deposit.updated",
  record: {
    tid:                      "TID9493F6CD41",
    uid:                      "ID092B2AF8E87",
    currency:                 "btc",
    amount:                   "0.0855",
    state:                    "accepted",
    created_at:               "2018-04-12T17:16:06+03:00",
    updated_at:               "2018-04-12T18:46:57+03:00",
    completed_at:             "2018-04-12T18:46:57+03:00",
    blockchain_address:       "n1Ytj6Hy57YpfueA2vtmnwJQs583bpYn7W",
    blockchain_txid:          "c37ae1677c4c989dbde9ac22be1f3ff3ac67ed24732a9fa8c9258fdff0232d72",
    blockchain_confirmations: 7
  },
  changes: {
    state:                    "submitted",
    completed_at:             nil,
    blockchain_confirmations: 1,
    updated_at:               "2018-04-12T17:16:06+03:00"
  }
}
```

| Field      | Description                                      |
| ---------- | ------------------------------------------------ |
| `record`   | The up-to-date deposit attributes.               |
| `changes`  | The changed deposit attributes and their values. |

## Format of `withdraw.created` event

```ruby
event: {
  name: "withdraw.created",
  record: {
    tid:             "TID892F29F094",
    uid:             "ID092B2AF8E87",
    rid:             "0xdA35deE8EDDeAA556e4c26268463e26FB91ff74f",
    currency:        "eth",
    amount:          "4.5485",
    fee:             "0.0015",
    state:           "prepared",
    created_at:      "2018-04-12T18:52:16+03:00",
    updated_at:      "2018-04-12T18:52:16+03:00",
    completed_at:    nil,
    blockchain_txid: nil
  }
}
```

| Field      | Description                          |
| ---------- | ------------------------------------ |
| `record`   | The up-to-date withdraw attributes.  |

## Format of `withdraw.updated` event

```ruby
event: {
  name: "withdraw.updated",
  record: {
    tid:             "TID892F29F094",
    uid:             "ID092B2AF8E87",
    rid:             "0xdA35deE8EDDeAA556e4c26268463e26FB91ff74f",
    currency:        "eth",
    amount:          "4.5485",
    fee:             "0.0015",
    state:           "succeed",
    created_at:      "2018-04-12T18:52:16+03:00",
    updated_at:      "2018-04-12T18:56:23+03:00",
    completed_at:    "2018-04-12T18:56:23+03:00",
    blockchain_txid: "0x9c34d1750e225a95938f9884e857ab6f55eedda43b159d13abf773fe6a916164"
  },
  changes: {
    state:           "processing",
    updated_at:      "2018-04-12T18:55:39+03:00",
    completed_at:    "2018-04-12T18:55:39+03:00",
    blockchain_txid: nil
  }
}
```

| Field      | Description                                      |
| ---------- | ------------------------------------------------ |
| `record`   | The up-to-date withdraw attributes.               |
| `changes`  | The changed withdraw attributes and their values. |

## Producing events using Ruby

```ruby
require "bunny"

def generate_jwt(jwt_payload)
  Kernel.abort "Please, see «Overview of RabbitMQ message» for implementation guide."
end

Bunny.run host: "localhost", port: 5672, username: "guest", password: "guest" do |session|
  channel     = session.channel
  exchange    = channel.direct("peatio.events")
  jwt_payload = {
    iss:  "peatio",
    jti:  SecureRandom.uuid,
    iat:  Time.now.to_i,
    exp:  Time.now.to_i + 60,
    data: {
      event: {
        name: "deposit.created",
        record: {
          tid:                      "TID9493F6CD41",
          uid:                      "ID092B2AF8E87",
          currency:                 "btc",
          amount:                   "0.0855",
          state:                    "submitted",
          created_at:               "2018-04-12T17:16:06+03:00",
          updated_at:               "2018-04-12T17:16:06+03:00",
          completed_at:             nil,
          blockchain_address:       "n1Ytj6Hy57YpfueA2vtmnwJQs583bpYn7W",
          blockchain_txid:          "c37ae1677c4c989dbde9ac22be1f3ff3ac67ed24732a9fa8c9258fdff0232d72",
          blockchain_confirmations: 1
        }      
      }
    }  
  }
  exchange.publish(generate_jwt(jwt_payload), routing_key: "deposit.created")
end
```

IMPORTANT: Don't forget to implement the logic for JWT exception handling!

## Producing events using `rabbitmqadmin`

`rabbitmqadmin publish routing_key=deposit.created payload=JWT exchange=peatio.events`

Don't forget to pass environment variable `JWT`.

## Consuming events using Ruby

```ruby
require "bunny"

def verify_jwt(jwt_payload)
  Kernel.abort "Please, see «Overview of RabbitMQ message» for implementation guide."
end

Bunny.run host: "localhost", port: 5672, username: "guest", password: "guest" do |session|
  channel  = session.channel
  exchange = channel.direct("peatio.events")
  queue    = channel.queue("", auto_delete: true, durable: true, exclusive: true)
                    .bind(exchange, routing_key: "deposit.update")
  queue.subscribe manual_ack: true, block: true do |delivery_info, metadata, payload|
    Kernel.puts verify_jwt(JSON.parse(payload)).fetch(:data)
    channel.ack(delivery_info.delivery_tag)
  rescue => e
    channel.nack(delivery_info.delivery_tag, false, true)
  end
end
```

IMPORTANT: Don't forget to implement the logic for JWT exception handling!

**Comment by yivo** [comment]
Closed via #914.
